aptos-move/framework/aptos-framework/sources/vesting_without_staking.move

```diff
@@ -0,0 +1,1174 @@
+        assert_active_vesting_contract(contract_address);
+
+        vest(contract_address);
+
+        let vesting_contract = borrow_global_mut<VestingContract>(contract_address);
+        verify_admin(admin, vesting_contract);
+
+        vesting_contract.state = VESTING_POOL_TERMINATED;
+
+        emit_event(
+            &mut vesting_contract.terminate_events,
+            TerminateEvent {
+                admin: vesting_contract.admin,
+                vesting_contract_address: contract_address,
+            },
+        );
+    }
+
+    /// Withdraw all funds to the preset vesting contract's withdrawal address. This can only be called if the contract
+    /// has already been terminated.
+    public entry fun admin_withdraw(admin: &signer, contract_address: address) acquires VestingContract {
+        let vesting_contract = borrow_global<VestingContract>(contract_address);
+        assert!(vesting_contract.state == VESTING_POOL_TERMINATED, error::invalid_state(EVESTING_CONTRACT_STILL_ACTIVE));
+
+        let vesting_contract = borrow_global_mut<VestingContract>(contract_address);
+        verify_admin(admin, vesting_contract);
+        let withdrawn_coins = coin::balance<AptosCoin>(contract_address);
+        let coins = coin::withdraw<AptosCoin>(admin, withdrawn_coins);
+        let amount = coin::value(&coins);
+        if (amount == 0) {
+            coin::destroy_zero(coins);
+            return
+        };
+        aptos_account::deposit_coins(vesting_contract.withdrawal_address, coins);
+
+        emit_event(
+            &mut vesting_contract.admin_withdraw_events,
+            AdminWithdrawEvent {
+                admin: vesting_contract.admin,
+                vesting_contract_address: contract_address,
+                amount,
+            },
+        );
+    }
+
+    public entry fun set_beneficiary(
+        admin: &signer,
+        contract_address: address,
+        shareholder: address,
+        new_beneficiary: address,
+    ) acquires VestingContract {
+        // Verify that the beneficiary account is set up to receive APT. This is a requirement so distribute() wouldn't
+        // fail and block all other accounts from receiving APT if one beneficiary is not registered.
+        assert_account_is_registered_for_apt(new_beneficiary);
+
+        let vesting_contract = borrow_global_mut<VestingContract>(contract_address);
+        verify_admin(admin, vesting_contract);
+
+        let old_beneficiary = get_beneficiary(vesting_contract, shareholder);
+        let beneficiaries = &mut vesting_contract.beneficiaries;
+        if (simple_map::contains_key(beneficiaries, &shareholder)) {
+            let beneficiary = simple_map::borrow_mut(beneficiaries, &shareholder);
+            *beneficiary = new_beneficiary;
+        } else {
+            simple_map::add(beneficiaries, shareholder, new_beneficiary);
+        };
```

Comment:
does `simple_map::upsert` work here?

Reply:
I think so, previously I am following aptos's approach and I think using `upsert` will make the code better.